### PR TITLE
feat: add SchedulerV3 

### PR DESCRIPTION
**Pull request by OlivierDehaene**
- Refactor code to allow supporting multiple versions of the generate.proto at the same time
- Add v3/generate.proto (ISO to generate.proto for now but allow for future changes without impacting v2 backends)
- Add Schedule trait to abstract queuing and batching mechanisms that will be different in the future
- Add SchedulerV2/V3 impl

### CODE REVIEW

File: router/src/server.rs [hunk 1]
```diff
@@ -1470,32 +1472,157 @@ pub async fn run(
...
+                    shard_info.requires_padding,
+                    shard_info.window_size,
+                    shard_info.speculate,
+                    generation_health,
+                ));
+                tracing::info!("Using scheduler V3");
+
+                (scheduler, health_ext, shard_info, max_batch_total_tokens)
+            }
+            Err(_) => {
```
Comment (Narsil) [inline router/src/server.rs:1560]: What kind of error does connecting a v3 client to a v2 shard create ?

File: proto/generate.proto [hunk 2]
```diff
@@ -51,27 +51,6 @@ message ClearCacheRequest {
...
-    }
-}
-
-message Input {
-    repeated InputChunk chunks = 1;
-  }
-
 enum GrammarType {
     GRAMMAR_TYPE_NONE = 0;
     GRAMMAR_TYPE_JSON = 1;
```
[the remark this answers is not in the document]
Comment (OlivierDehaene) [reply proto/generate.proto:56]: No I added it to v3 directly to avoid any changes to v2.

File: router/src/server.rs (hunk 1 above)
Comment (OlivierDehaene) [reply router/src/server.rs:1560]: `ClientError::Connection("Server does not support v3 interface".to_string())`

File: proto/generate.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (Narsil) [reply proto/generate.proto:56]: I meant the file generate.proto

**Review by Narsil — APPROVED** [review]
LGTM

**Pull request merged**